python/ray/WebUI.ipynb

```diff
@@ -1,61 +1,194 @@
 {
  "cells": [
   {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "# Evaluate the box below to initialize the web UI."
-   ]
-  },
-  {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "collapsed": true
+   },
    "outputs": [],
    "source": [
+    "import ipywidgets as widgets\n",
     "import os\n",
+    "import pandas as pd\n",
     "import pprint\n",
-    "import ipywidgets as widgets\n",
+    "import qgrid\n",
     "import ray\n",
+    "import subprocess\n",
+    "import sys\n",
+    "import tempfile\n",
+    "import time\n",
+    "import math\n",
     "\n",
     "from IPython.display import display\n",
     "\n",
-    "ray.init(redis_address=os.environ[\"REDIS_ADDRESS\"])"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "**Evaluate the box below to search for objects.**"
+    "ray.init(redis_address=os.environ[\"REDIS_ADDRESS\"])\n"
    ]
   },
   {
    "cell_type": "code",
    "execution_count": null,
-   "metadata": {},
+   "metadata": {
+    "collapsed": true
+   },
    "outputs": [],
    "source": [
-    "object_search = widgets.Text(\n",
-    "    value=\"\",\n",
-    "    placeholder=\"Object ID\",\n",
-    "    description=\"Search for an object:\",\n",
-    "    disabled=False\n",
-    ")\n",
-    "display(object_search)\n",
+    "class _EventRecursionContextManager(object):\n",
```

Comment:
There's a single instance of that recursion manager class, out_recursion, whose purpose it is to store whether or not someone is currently in the callback method, as a crude but probably correct way or figuring out whether the next call to the method would be from a human or from code. The recursion manager class implements two methods, __enter__ and __exit__, that allow it to be used as a context manager, so when you have a block annotated with "with out_recursion", when it enters that block it calls out_recursion.__enter__(), which sets some "currently executing" flag to true, and any time you exit, whether through a return or an exception or anything, it calls the out_recursion.__exit__() method, which sets that "currently executing" flag to false. So effectively when a person changes the slider, nothing is currently in the callback, so the slider callback starts to execute, which means setting the flag to true, then the slider sets the value on the text boxes, which tries to execute their callbacks but notices the flag is set to the true so they immediate exit, then the slider callback finishes doing whatever and returns, which sets the flag back to false for when the person next changes anything.